HttpJson UpdateUserRequest's FieldMask is unable to be Unmarshalled

Java Gapic-Generator is implementing showcase coverage for the Identity client (modeled after golang). Draft PR with the full sample code: https://github.com/googleapis/gapic-generator-java/pull/1431 

Relevant Code for the test case:
```
CreateUserRequest createUserRequest = CreateUserRequest.newBuilder()
            .setUser(User.newBuilder().setDisplayName("Jane Doe").setEmail("janedoe@example.com").setNickname("Doe").setHeightFeet(5).build()).build();
    User user = httpjsonClient.createUser(createUserRequest);
    User expected = createUserRequest.getUser();

    UpdateUserRequest updateUserRequest = UpdateUserRequest.newBuilder()
            .setUser(User.newBuilder()
                    .setName(user.getName())
                    .setDisplayName(user.getDisplayName())
                    .setEmail("janedoe@jane.com")
                    .setHeightFeet(6.0)
                    .setEnableNotifications(true)
                    .build())
            .setUpdateMask(FieldMask.newBuilder().addAllPaths(Arrays.asList("email", "height_feet", "enable_notifications")).build())
            .build();
    User updatedUser = httpjsonClient.updateUser(updateUserRequest);
```

Results in this error:
```
Caused by: com.google.api.client.http.HttpResponseException: 400 Bad Request
POST http://localhost:7469/v1beta1/users/1?updateMask=email,heightFeet,enableNotifications
{"error":{"code":400,"message":"error reading query params: terminal field \"updateMask\" of field path \"updateMask\" is of type \"message\" with value string \"email,heightFeet,enableNotifications\", which could not be parsed: unable to unmarshal field \"update_mask\" in message \"google.showcase.v1beta1.UpdateUserRequest\": proto: syntax error (line 1:1): invalid value email","details":null,"Body":"","Header":null,"Errors":null}}
```

Logging the HttpRequest:
```
-------------- REQUEST  --------------
POST http://localhost:7469/v1beta1/users/1?updateMask=email,heightFeet,enableNotifications
Accept-Encoding: gzip
User-Agent: Google-HTTP-Java-Client/1.43.0 (gzip)
x-http-method-override: PATCH
x-goog-api-client: gl-java/11.0.16.1 gapic/0.0.1-SHAPSHOT gax/2.23.3-SNAPSHOT rest/
Content-Type: application/json; charset=utf-8
Content-Length: 115

curl -v --compressed -X POST -H 'Accept-Encoding: gzip' -H 'User-Agent: Google-HTTP-Java-Client/1.43.0 (gzip)' -H 'x-http-method-override: PATCH' -H 'x-goog-api-client: gl-java/11.0.16.1 gapic/0.0.1-SHAPSHOT gax/2.23.3-SNAPSHOT rest/' -H 'Content-Type: application/json; charset=utf-8' -d '@-' -- 'http://localhost:7469/v1beta1/users/12?updateMask=email,heightFeet,enableNotifications' << $$$
Total: 115 bytes
{"name":"users/12","displayName":"Jane Doe","email":"janedoe@jane.com","heightFeet":6.0,"enableNotifications":true}
```
The logs seems to show that the updateMask field is encodedproperly as a query param, but Showcase is unable to decode this query param:
```
2023/03/07 10:16:40 error reading query params: terminal field "updateMask" of field path "updateMask" is of type "message" with value string "email,heightFeet,enableNotifications", which could not be parsed: unable to unmarshal field "update_mask" in message "google.showcase.v1beta1.UpdateUserRequest": proto: syntax error (line 1:1): invalid value email
```

---
Thought this might be an error with Java's HttpJson implementation for FieldMask, but tested this with Java-Scheduler using HttpJson + FieldMask and this works properly. 

```
-------------- REQUEST  --------------
POST https://cloudscheduler.googleapis.com:443/v1/projects/lawrence-test-project-2/locations/us-central1/jobs/72824232043972625121?$alt=json;enum-encoding%3Dint&updateMask=schedule,timeZone
Accept-Encoding: gzip
Authorization: Bearer {BEARER}
User-Agent: Google-HTTP-Java-Client/1.42.3 (gzip)
x-goog-user-project: {PROJECT}
x-http-method-override: PATCH
x-goog-api-client: gl-java/11.0.16.1 gapic/2.11.0 gax/2.23.1 rest/
Content-Type: application/json
Content-Length: 330

curl -v --compressed -X POST -H 'Accept-Encoding: gzip' -H 'Authorization: Bearer {BEARER}' -H 'User-Agent: Google-HTTP-Java-Client/1.42.3 (gzip)' -H 'x-goog-user-project: lawrence-test-project-2' -H 'x-http-method-override: PATCH' -H 'x-goog-api-client: gl-java/11.0.16.1 gapic/2.11.0 gax/2.23.1 rest/' -H 'Content-Type: application/json' -d '@-' -- 'https://cloudscheduler.googleapis.com:443/v1/projects/{PROJECT}/locations/{LOCATION}/jobs/{ID}?$alt=json;enum-encoding%3Dint&updateMask=schedule,timeZone' << $$$
Total: 330 bytes
{"name":"projects/{PROJECT}/locations/{LOCATION}/jobs/{ID}","httpTarget":{"uri":"http://example.com/","httpMethod":2,"headers":{"User-Agent":"Google-Cloud-Scheduler"}},"userUpdateTime":"2023-03-07T15:57:34Z","state":1,"schedule":"1 10 * * FRI","timeZone":"America/New_York","attemptDeadline":"180s"}
```

--- a/util/genrest/goviewcreator.go
+++ b/util/genrest/goviewcreator.go
@@ -75,7 +75,7 @@ func NewView(model *gomodel.Model) (*goview.View, error) {
 
 			source.P("")
 			source.P("// %s translates REST requests/responses on the wire to internal proto messages for %s", handlerName, handler.GoMethod)
-			source.P("//    Generated for HTTP binding pattern: %q", handler.URIPattern)
+			source.P("//    Generated for HTTP binding pattern: %s %q", handler.HTTPMethod, handler.URIPattern)
 			source.P("func (backend *RESTBackend) %s(w http.ResponseWriter, r *http.Request) {", handlerName)
 			if handler.StreamingClient {
 				source.P(`  backend.Error(w, http.StatusNotImplemented, "client-streaming methods not implemented yet (request matched '%s': %%q)", r.URL)`, handler.URIPattern)

--- a/util/genrest/resttools/headers.go
+++ b/util/genrest/resttools/headers.go
@@ -56,7 +56,8 @@ func CheckContentType(header http.Header) error {
 func CheckAPIClientHeader(header http.Header) error {
 	content, ok := header[headerNameAPIClient]
 	if !ok || len(content) != 1 {
-		return fmt.Errorf("(HeaderAPIClientError) did not find expected HTTP header %q: %q", headerNameAPIClient, headerValueTransportRESTPrefix)
+		return fmt.Errorf("(HeaderAPIClientError) did not find expected HTTP header %q: %q\n                found: %q",
+			headerNameAPIClient, headerValueTransportRESTPrefix, header)
 	}
 
 	var haveREST, haveGAPIC bool

--- a/util/genrest/resttools/populatefield.go
+++ b/util/genrest/resttools/populatefield.go
@@ -268,7 +268,7 @@ func parseWellKnownType(message protoreflect.Message, fieldDescriptor protorefle
 	}
 
 	if stringEncoded {
-		value = fmt.Sprintf(`"%s"`, value)
+		value = fmt.Sprintf("%q", value)
 	}
 
 	msgValue := fieldMsg.New()

--- a/util/genrest/resttools/populatefield.go
+++ b/util/genrest/resttools/populatefield.go
@@ -24,43 +24,31 @@ import (
 	"google.golang.org/protobuf/reflect/protoreflect"
 )
 
-type valueTransform func(string) (string, error)
-
-func ensureQuotedValue(pre string) (string, error) {
-	hasStartQuote := strings.HasPrefix(pre, `"`)
-	hasEndQuote := strings.HasSuffix(pre, `"`)
-	if hasStartQuote != hasEndQuote {
-		return "", fmt.Errorf("unbalanced quotes in value %q", pre)
-	}
-	if hasStartQuote {
-		return pre, nil
-	}
-	return fmt.Sprintf(`"%s"`, pre), nil
-}
-
-var wellKnownTypes = map[string]valueTransform{
-	// testing various types in https://go.dev/play/p/rNhsXY364qQ
-
+// wellKnownTypes has a key for every well-known type message which JSON-encodes to an atomic symbol
+// (like a number or a string) as opposed to a structured object. The value for each key is true iff
+// the JSON encoding for the type is a string. We need both these data for well-known types so that
+// we can properly decode them in URL paths and query strings.
+var wellKnownTypes = map[string]bool{
 	// == The following are the only three common types used in this API ==
-	"google.protobuf.FieldMask": ensureQuotedValue,
-	"google.protobuf.Timestamp": ensureQuotedValue,
-	"google.protobuf.Duration":  ensureQuotedValue,
+	"google.protobuf.FieldMask": true,
+	"google.protobuf.Timestamp": true,
+	"google.protobuf.Duration":  true,
 	// == End utilized types ==
-	// TODO: When the following start being used int he Showcase API, add tests for their use as
-	// (unquoted) query params. These types are defined in
+	// TODO: When the following start being used in the Showcase API, add tests for each of
+	// these. These types are defined in
 	// https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/wrappers.proto
-	"google.protobuf.DoubleValue": nil,
-	"google.protobuf.FloatValue":  nil,
-	"google.protobuf.Int64Value":  ensureQuotedValue,
-	"google.protobuf.UInt64Value": ensureQuotedValue,
-	"google.protobuf.Int32Value":  nil,
-	"google.protobuf.UInt32Value": nil,
-	"google.protobuf.BoolValue":   nil,
-	"google.protobuf.StringValue": ensureQuotedValue,
-	"google.protobuf.BytesValue":  ensureQuotedValue,
+	"google.protobuf.DoubleValue": false,
+	"google.protobuf.FloatValue":  false,
+	"google.protobuf.Int64Value":  true,
+	"google.protobuf.UInt64Value": true,
+	"google.protobuf.Int32Value":  false,
+	"google.protobuf.UInt32Value": false,
+	"google.protobuf.BoolValue":   false,
+	"google.protobuf.StringValue": true,
+	"google.protobuf.BytesValue":  true,
 	// TODO: Determine if the following are even viable as query params.
-	"google.protobuf.Value":     nil,
-	"google.protobuf.ListValue": nil,
+	"google.protobuf.Value":     false,
+	"google.protobuf.ListValue": false,
 }
 
 // PopulateSingularFields sets the fields within protoMessage to the values provided in
@@ -153,6 +141,12 @@ func PopulateOneField(protoMessage proto.Message, fieldPath string, fieldValues 
 		kind := fieldDescriptor.Kind()
 		switch kind {
 		case protoreflect.MessageKind:
+			// The only `MessageKind`s we accept in URL paths and query params are
+			// well-known types where the message as a whole encodes into a format
+			// similar to a regular terminal field. Normal messages conveyed via URL
+			// paths or query params must be non-repeated and are represented by listing
+			// their terminal primitive fields, which will trigger the other cases
+			// below instead of this one.
 			if pval, err := parseWellKnownType(message, fieldDescriptor, value); err != nil {
 				parseError = err
 			} else if pval != nil {
@@ -268,17 +262,13 @@ func parseWellKnownType(message protoreflect.Message, fieldDescriptor protorefle
 	}
 	fieldMsg := messageFieldTypes.Message(fieldDescriptor.Index())
 	fullName := string(fieldMsg.Descriptor().FullName())
-	transform, isWellKnown := wellKnownTypes[fullName]
+	stringEncoded, isWellKnown := wellKnownTypes[fullName]
 	if !isWellKnown {
 		return nil, nil
 	}
 
-	if transform != nil {
-		var err error
-		if value, err = transform(value); err != nil {
-			return nil, err
-		}
-
+	if stringEncoded {
+		value = fmt.Sprintf(`"%s"`, value)
 	}
 
 	msgValue := fieldMsg.New()

--- a/util/genrest/resttools/populatefield.go
+++ b/util/genrest/resttools/populatefield.go
@@ -24,24 +24,43 @@ import (
 	"google.golang.org/protobuf/reflect/protoreflect"
 )
 
-var wellKnownTypes = map[string]bool{
+type valueTransform func(string) (string, error)
+
+func ensureQuotedValue(pre string) (string, error) {
+	hasStartQuote := strings.HasPrefix(pre, `"`)
+	hasEndQuote := strings.HasSuffix(pre, `"`)
+	if hasStartQuote != hasEndQuote {
+		return "", fmt.Errorf("unbalanced quotes in value %q", pre)
+	}
+	if hasStartQuote {
+		return pre, nil
+	}
+	return fmt.Sprintf(`"%s"`, pre), nil
+}
+
+var wellKnownTypes = map[string]valueTransform{
+	// testing various types in https://go.dev/play/p/rNhsXY364qQ
+
 	// == The following are the only three common types used in this API ==
-	"google.protobuf.FieldMask": true,
-	"google.protobuf.Timestamp": true,
-	"google.protobuf.Duration":  true,
+	"google.protobuf.FieldMask": ensureQuotedValue,
+	"google.protobuf.Timestamp": ensureQuotedValue,
+	"google.protobuf.Duration":  ensureQuotedValue,
 	// == End utilized types ==
-	"google.protobuf.DoubleValue": true,
-	"google.protobuf.FloatValue":  true,
-	"google.protobuf.Int64Value":  true,
-	"google.protobuf.UInt64Value": true,
-	"google.protobuf.Int32Value":  true,
-	"google.protobuf.UInt32Value": true,
-	"google.protobuf.BoolValue":   true,
-	"google.protobuf.StringValue": true,
-	"google.protobuf.BytesValue":  true,
+	// TODO: When the following start being used int he Showcase API, add tests for their use as
+	// (unquoted) query params. These types are defined in
+	// https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/wrappers.proto
+	"google.protobuf.DoubleValue": nil,
+	"google.protobuf.FloatValue":  nil,
+	"google.protobuf.Int64Value":  ensureQuotedValue,
+	"google.protobuf.UInt64Value": ensureQuotedValue,
+	"google.protobuf.Int32Value":  nil,
+	"google.protobuf.UInt32Value": nil,
+	"google.protobuf.BoolValue":   nil,
+	"google.protobuf.StringValue": ensureQuotedValue,
+	"google.protobuf.BytesValue":  ensureQuotedValue,
 	// TODO: Determine if the following are even viable as query params.
-	"google.protobuf.Value":     true,
-	"google.protobuf.ListValue": true,
+	"google.protobuf.Value":     nil,
+	"google.protobuf.ListValue": nil,
 }
 
 // PopulateSingularFields sets the fields within protoMessage to the values provided in
@@ -249,8 +268,17 @@ func parseWellKnownType(message protoreflect.Message, fieldDescriptor protorefle
 	}
 	fieldMsg := messageFieldTypes.Message(fieldDescriptor.Index())
 	fullName := string(fieldMsg.Descriptor().FullName())
-	if !wellKnownTypes[fullName] {
+	transform, isWellKnown := wellKnownTypes[fullName]
+	if !isWellKnown {
 		return nil, nil
+	}
+
+	if transform != nil {
+		var err error
+		if value, err = transform(value); err != nil {
+			return nil, err
+		}
+
 	}
 
 	msgValue := fieldMsg.New()

--- a/util/genrest/resttools/populatefield_test.go
+++ b/util/genrest/resttools/populatefield_test.go
@@ -31,6 +31,8 @@ import (
 )
 
 func TestParseWellKnownType(t *testing.T) {
+	// TODO: Test the other well-known types as the Showcase API
+	// starts incorporating them.
 	for _, tst := range []struct {
 		name  string
 		msg   protoreflect.Message
@@ -62,14 +64,67 @@ func TestParseWellKnownType(t *testing.T) {
 
 		gotp, err := parseWellKnownType(tst.msg, fd, value)
 		if err != nil {
-			t.Fatal(err)
+			t.Fatalf("parsing %q led to error %s", value, err)
 		}
 		if gotp == nil {
-			t.Fatal("expected non-nil value from parsing")
+			t.Fatalf("expected non-nil value from parsing: %s", value)
 		}
 		got := gotp.Message().Interface()
 		if diff := cmp.Diff(got, tst.want, cmp.Comparer(proto.Equal)); diff != "" {
 			t.Fatalf("%s: got(-),want(+):\n%s", "FieldMask", diff)
+		}
+	}
+}
+
+func TestParseWellKnownTypeUnquoted(t *testing.T) {
+	// TODO: Test the other well-known types as the Showcase API
+	// starts incorporating them.
+	for _, tst := range []struct {
+		name  string
+		msg   protoreflect.Message
+		field protoreflect.Name
+		value string
+		want  proto.Message
+	}{
+		{
+			"google.protobuf.FieldMask",
+			(&genprotopb.UpdateUserRequest{}).ProtoReflect(),
+			"update_mask",
+			`foo,bar,baz`,
+			&fieldmaskpb.FieldMask{Paths: []string{"foo", "bar", "baz"}},
+		},
+
+		{
+			"google.protobuf.Timestamp",
+			(&genprotopb.User{}).ProtoReflect(),
+			"create_time",
+			"2023-03-21T12:01:02.000000003Z",
+			timestamppb.New(time.Date(2023, 03, 21, 12, 1, 2, 3, time.UTC)),
+		},
+
+		{
+			"google.protobuf.Duration",
+			(&genprotopb.Sequence_Response{}).ProtoReflect(),
+			"delay",
+			"5s",
+			durationpb.New(5 * time.Second),
+		},
+	} {
+		fd := tst.msg.Descriptor().Fields().ByName(tst.field)
+
+		gotp, err := parseWellKnownType(tst.msg, fd, tst.value)
+		if err != nil {
+			t.Errorf("parsing %q led to error %s", tst.value, err)
+			continue
+		}
+		if gotp == nil {
+			t.Errorf("expected non-nil value from parsing: %s", tst.value)
+			continue
+		}
+		got := gotp.Message().Interface()
+		if diff := cmp.Diff(got, tst.want, cmp.Comparer(proto.Equal)); diff != "" {
+			t.Errorf("%s: got(-),want(+):\n%s", "FieldMask", diff)
+			continue
 		}
 	}
 }

--- a/util/genrest/resttools/populatefield_test.go
+++ b/util/genrest/resttools/populatefield_test.go
@@ -21,7 +21,6 @@ import (
 
 	"github.com/google/go-cmp/cmp"
 	genprotopb "github.com/googleapis/gapic-showcase/server/genproto"
-	"google.golang.org/protobuf/encoding/protojson"
 	"google.golang.org/protobuf/encoding/prototext"
 	"google.golang.org/protobuf/proto"
 	"google.golang.org/protobuf/reflect/protoreflect"
@@ -37,52 +36,6 @@ func TestParseWellKnownType(t *testing.T) {
 		name  string
 		msg   protoreflect.Message
 		field protoreflect.Name
-		want  proto.Message
-	}{
-		{
-			"google.protobuf.FieldMask",
-			(&genprotopb.UpdateUserRequest{}).ProtoReflect(),
-			"update_mask",
-			&fieldmaskpb.FieldMask{Paths: []string{"foo", "bar", "baz"}},
-		},
-		{
-			"google.protobuf.Timestamp",
-			(&genprotopb.User{}).ProtoReflect(),
-			"create_time",
-			timestamppb.Now(),
-		},
-		{
-			"google.protobuf.Duration",
-			(&genprotopb.Sequence_Response{}).ProtoReflect(),
-			"delay",
-			durationpb.New(5 * time.Second),
-		},
-	} {
-		data, _ := protojson.Marshal(tst.want)
-		value := string(data)
-		fd := tst.msg.Descriptor().Fields().ByName(tst.field)
-
-		gotp, err := parseWellKnownType(tst.msg, fd, value)
-		if err != nil {
-			t.Fatalf("parsing %q led to error %s", value, err)
-		}
-		if gotp == nil {
-			t.Fatalf("expected non-nil value from parsing: %s", value)
-		}
-		got := gotp.Message().Interface()
-		if diff := cmp.Diff(got, tst.want, cmp.Comparer(proto.Equal)); diff != "" {
-			t.Fatalf("%s: got(-),want(+):\n%s", "FieldMask", diff)
-		}
-	}
-}
-
-func TestParseWellKnownTypeUnquoted(t *testing.T) {
-	// TODO: Test the other well-known types as the Showcase API
-	// starts incorporating them.
-	for _, tst := range []struct {
-		name  string
-		msg   protoreflect.Message
-		field protoreflect.Name
 		value string
 		want  proto.Message
 	}{
@@ -90,7 +43,7 @@ func TestParseWellKnownTypeUnquoted(t *testing.T) {
 			"google.protobuf.FieldMask",
 			(&genprotopb.UpdateUserRequest{}).ProtoReflect(),
 			"update_mask",
-			`foo,bar,baz`,
+			"foo,bar,baz",
 			&fieldmaskpb.FieldMask{Paths: []string{"foo", "bar", "baz"}},
 		},
 

